Add doNotStartTrace Option

johnbelamaric opened the pull request:
The is option will only trace gRPC calls if the Context already
contains a Span. This allows the caller to choose if any given
request is traced or not.

johnbelamaric committed:
Add doNotStartTrace Option

The is option will only trace gRPC calls if the Context already
contains a Span. This allows the caller to choose if any given
request is traced or not.

bhs commented on go/otgrpc/options.go:
```diff
@@ -17,6 +17,14 @@ func LogPayloads() Option {
 
+// DoNotStartTrace returns an Option that tells the OpenTracing instrumentation to
+// only trace if the context already contains a Span; do not start a new Span
+func DoNotStartTrace() Option {
```
can we make this a little more generic? something like

```
// Returns true if.f. a Span should be created.
type SpanFilter(parentSpanCtx opentracing.SpanContext, method string, req, resp interface{}) bool
```

... then follow the pattern for the SpanDecorator. Seems more general?

johnbelamaric replied:
Sure, good idea.

johnbelamaric committed:
Implement as SpanFilter

Change the implementation to use a SpanFilterFunc, allowing the
user to provide a function to determine whether or not to start
a span.

johnbelamaric commented:
@bhs ok, PTAL

johnbelamaric commented:
I signed it!

bhs commented on go/otgrpc/options.go:
```diff
@@ -17,6 +17,21 @@ func LogPayloads() Option {
+// SpanFilterFunc provides an optional mechanism to decided whether or not
+// to trace a given gRPC call. Return true to create a Span and initiate
+// tracing, false to not create a Span and not trace
+type SpanFilterFunc func(
```
sorry, one last nitpick: let's call it `IncludeSpanFunc`... some people think "filter" means keep, others think "filter" means discard.

bhs commented on go/otgrpc/options.go:
```diff
@@ -17,6 +17,21 @@ func LogPayloads() Option {
 	}
 }
 
+// SpanFilterFunc provides an optional mechanism to decided whether or not
```
nit: "to decide"

bhs reviewed (commented):
Last round of nits, then this LGTM! Thanks, this is a nice improvement.

johnbelamaric committed:
Rename Filter to Include

johnbelamaric commented:
Ok, nits resolved.

bhs reviewed (approved):
Thanks. I have a few tiny nits, but I will do them myself – no need to trouble you!

johnbelamaric commented:
@bhs thanks!
